Выполнить ТЗ спринта №7

SergeyLeshchev opened the pull request:
1. Создать класс FileBackedTaskManager - менеджер, сохраняющий задачи в файл
2. Сделать в InMemoryTaskManager некоторые поля и методы теперь не приватными
3. Создать класс исключения ManagerSaveException
4. Создать множество TaskType для типов задач
5. Создать тестовый класс FileBackedTaskManagerTest

Commit:
Выполнить ТЗ спринта №7

1. Создать класс FileBackedTaskManager - менеджер, сохраняющий задачи в файл
2. Сделать в InMemoryTaskManager некоторые поля и методы теперь не приватными
3. Создать класс исключения ManagerSaveException
4. Создать множество TaskType для типов задач
5. Создать тестовый класс FileBackedTaskManagerTest

Commit:
Выполнить дополнение к ТЗ спринта №7

1. Исправить InMemoryTaskManager, чтобы снова правильно функционировала история
2. Вернуть пользовательский сценарий с историей в класс Main

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -10,11 +10,11 @@
-    private final HashMap<Integer, Epic> epicCollection = new HashMap<>();
-    private final HashMap<Integer, Subtask> subtaskCollection = new HashMap<>();
-    private int idOfTasks = 1;
+    final HistoryManager historyManager = Managers.getDefaultHistoryManager();
+    final HashMap<Integer, Task> taskCollection = new HashMap<>();
+    final HashMap<Integer, Epic> epicCollection = new HashMap<>();
+    final HashMap<Integer, Subtask> subtaskCollection = new HashMap<>();
+    int idOfTasks = 1;
```
Не стоит увеличивать область видимости выше protected, с protected поля будут доступны в наследниках

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
```
`implements TaskManager ` не нужно, `InMemoryTaskManager` реализует этот интерфейс

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+        save();
+    }
+
+    void calculateEpicStatus(Epic epic) {
+        super.calculateEpicStatus(epic);
+        save();
+    }
```
Этот метод переопределять не нужно, он вшит в логику нужных методов в `InMemoryTaskManager`

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+        return this.savedTasks;
+    }
+
+    private static String toStringTask(Task task) {
+        String taskString = task.getIdOfTask() + ",";
+        if (task.getClass().toString().endsWith("Task")) {
+            taskString += "TASK,";
+        } else if (task.getClass().toString().endsWith("Epic")) {
+            taskString += "EPIC,";
+        } else {
+            taskString += "SUBTASK,";
+        }
+        taskString += task.getTitle() + "," + task.getStatus() + "," + task.getDescription() + ",";
+        if (task.getClass().toString().endsWith("Subtask")) {
+            Subtask subtask = (Subtask) task;
+            taskString += subtask.getEpicId();
+        }
+        return taskString;
+    }
+
+    private Task fromString(String value) {
+        String[] taskArray = value.split(",");
+        Task task = null;
+        switch (taskArray[1]) {
+            case "TASK" -> task = new Task(taskArray[2], taskArray[4], Status.valueOf(taskArray[3]),
+                        Integer.parseInt(taskArray[0]));
+            case "EPIC" -> task = new Epic(taskArray[2], taskArray[4], Status.valueOf(taskArray[3]),
+                        Integer.parseInt(taskArray[0]));
+            case "SUBTASK" -> task = new Subtask(taskArray[2], taskArray[4], Status.valueOf(taskArray[3]),
+                        Integer.parseInt(taskArray[5]), Integer.parseInt(taskArray[0]));
+        }
+        return task;
+    }
```
Будет лучше вынести методы в утилитный класс, чтобы разгрузить менеджер

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+        String[] taskArray = value.split(",");
+        Task task = null;
+        switch (taskArray[1]) {
+            case "TASK" -> task = new Task(taskArray[2], taskArray[4], Status.valueOf(taskArray[3]),
+                        Integer.parseInt(taskArray[0]));
+            case "EPIC" -> task = new Epic(taskArray[2], taskArray[4], Status.valueOf(taskArray[3]),
+                        Integer.parseInt(taskArray[0]));
+            case "SUBTASK" -> task = new Subtask(taskArray[2], taskArray[4], Status.valueOf(taskArray[3]),
```
В кейсах используй енамы `TaskType`
Распарсить строку и записать значения по переменным тоже стоит до свитч кейса, чтобы не дублировать в кейсах и не выглядело сложно)

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+                if (fileBackedTaskManager.idOfTasks < task.getIdOfTask()) {
+                    fileBackedTaskManager.idOfTasks = task.getIdOfTask();
+                }
+                if (task.getClass().toString().endsWith("Task")) {
+                    fileBackedTaskManager.taskCollection.put(task.getIdOfTask(), task);
+                } else if (task.getClass().toString().endsWith("Epic")) {
+                    fileBackedTaskManager.epicCollection.put(task.getIdOfTask(), (Epic) task);
+                } else {
```
Используй енамы, иначе какой в них смысл, добавь геттеры в классы задач, которые будут возвращать тип в зависимости от класса, где он реализован.

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,198 @@
+        this.savedTasks = savedTasks;
+    }
+
+    public File getSavedTasks() {
+        return this.savedTasks;
+    }
```
Этот геттер не стоит предоставлять, менеджер сам работает с файлом

SergeyLeshchev replied:
После того, как я убрал этот геттер, в тестовом классе появились ошибки, так как там использовался этот геттер. Я понял, что я могу использовать ссылку на файл напрямую, а не через taskManager, и заменил ссылки там, где использовался геттер. Ошибок нет, но теперь не совсем понятно, что тестирует этот метод, зачем и нужен ли вообще? В ТЗ написано, что нужно проверить сохранение и загрузку пустого файла, этот тест выполняет эту задачу или я делаю что-то неправильно? Я могу обратиться к наставнику, если требуются длительные объяснения)
`    @Test
    public void emptyFileLoadingTest() {
        try {
            FileBackedTaskManager taskManager = FileBackedTaskManager.loadFromFile(new File(testFile.getName()));
            assertNotNull(testFile, "file savedTasks is null");
        } catch (IOException e) {
            System.out.println(e.getMessage());
            throw new ManagerSaveException("Неправильно работает тест на загрузку пустого файла");
        }
    }`

Ivanov-s-tmn commented on test/manager/FileBackedTaskManagerTest.java:
Вынеси создание файла, файлового менеджера в метод BeforeEach, а удаление тестового файла - в метод AfterEach. Тогда никакие обработки исключений не нужны. В тестах работай только с методами менеджера для проверки его работы, примерный тест кейс:
1. Создать файловый менеджер
2. Заполнить тестовыми данными
3. Создать новый файловый менеджер из этого же файла через loadFromFile()
4. Проверить, что новый менеджер корректно восстановил состояние из файла (списки задач до и после выгрузки должны быть идентичны)

Этот один тест будет проверять как запись в файл, так и восстановление данных из файла

Commit:
Выполнить работу над ошибками ТЗ спринта №7

1. Сделать некоторые поля и методы protected и private
2. Удалить метод calculateEpicStatus из FileBackedTaskManager
3. Переместить методы toStringTask fromString в класс TaskSerializer
4. Использовать TypeTask в методах TaskSerializer
5. Добавить getType в Task, Epic и Subtask
6. Удалить геттер getSavedTasks в FileBackedTaskManager
7. Изменить тестовый класс FileBackedTaskManagerTest
8. Удалить лишние сеттеры в классах Task и Epic

Ivanov-s-tmn commented on test/manager/FileBackedTaskManagerTest.java:
```diff
@@ -5,107 +5,55 @@
-        } finally {
-            assert testFile != null;
-            testFile.delete();
+            throw new ManagerSaveException("Неправильно работает тестовый метод saveAndLoadTasksTest");
```
В тестах не нужна обработка исключений, в методе `loadFromFile()` исключение `IOException` обрабатывается и выбрасывается `ManagerSaveException`, соответственно тут уже никогда не возникнет исключения `IOException` и блок try catch бесполезен

SergeyLeshchev replied:
Ага, теперь понял, спасибо)

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -42,47 +42,16 @@ public FileBackedTaskManager(File savedTasks) throws IOException {
-        }
-        return task;
-    }
+
+
```
Лишние отступы

Ivanov-s-tmn commented on src/manager/TaskSerializer.java:
```diff
@@ -0,0 +1,38 @@
+
+public class TaskSerializer {
+    public static String toStringTask(Task task) {
+        String taskString = task.getIdOfTask() + ",";
+        switch (task.getType()) {
+            case TASK -> taskString += "TASK,";
+            case EPIC -> taskString += "EPIC,";
+            case SUBTASK -> taskString += "SUBTASK,";
+        }
+        taskString += task.getTitle() + "," + task.getStatus() + "," + task.getDescription() + ",";
+        if (task.getType().equals(TaskType.SUBTASK)) {
+            Subtask subtask = (Subtask) task;
+            taskString += subtask.getEpicId();
+        }
+        return taskString;
```
При частой конкатенации стоит использовать `StringBuilder`, либо вообще можно переопределить методы `toString()` в классах задач и использовать их для записи в файл

SergeyLeshchev replied:
Переопределять родной toString я бы не очень хотел, потому что мне больше нравится его текущая реализация, которую я использую при отладке и в пользовательских сценариях. Например сейчас toString показывает список сабтасков в эпике. И в будущем, если бы кто-то работал с кодом после меня, то поведение этого toString более предсказуемое. Про StringBuilder спасибо

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -42,47 +42,16 @@ public FileBackedTaskManager(File savedTasks) throws IOException {
+
+
 
     public static FileBackedTaskManager loadFromFile(File file) throws IOException {
```
`throws IOException` не нужен, это исключение обрабатывается.
В конструкторе аналогично

Commit:
Выполнить работу над ошибками №2 в ТЗ спринта №7

1. Удалить try-catch в main в FileBackedTaskManager
2. Заменить if-else на switch в loadFromFile в FileBackedTaskManager
3. Заменить String на StringBuilder в toStringTask в TaskSerializer
4. Удалить try-catch в saveAndLoadTasksTest в FileBackedTaskManagerTest
5. Внести другие незначительные исправления

Commit:
Выполнить работу над ошибками №2 в ТЗ спринта №7

1. Удалить try-catch в main в FileBackedTaskManager
2. Заменить if-else на switch в loadFromFile в FileBackedTaskManager
3. Заменить String на StringBuilder в toStringTask в TaskSerializer
4. Удалить try-catch в saveAndLoadTasksTest в FileBackedTaskManagerTest
5. Внести другие незначительные исправления

Commit:
Выполнить работу над ошибками №2 в ТЗ спринта №7

1. Исправить ошибки codestyle
